Stops parsing after first token if its size is very close to the size of the parser's buffer

As detected by the fuzz test, the parser stops parsing the stream after the first token if its size is very close to the size of the parser's buffer:

> There are also a few other issues after fixing this one:
> 
> ```console
> $ go test -fuzz .
> fuzz: elapsed: 0s, gathering baseline coverage: 0/106 completed
> fuzz: elapsed: 0s, gathering baseline coverage: 106/106 completed, now fuzzing with 32 workers
> fuzz: elapsed: 1s, execs: 71983 (48607/sec), new interesting: 21 (total: 127)
> --- FAIL: FuzzStream (1.48s)
>     --- FAIL: FuzzStream (0.00s)
>         tokenizer_test.go:397: {
>             	Id: 0
>             	Key: -1
>             	Value: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
>             	Position: 0
>             	Indent: 0 bytes
>             	Line: 1
>             }
>         tokenizer_test.go:411: input:
>             "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA0000"
>             expected:
>             "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA0000"
>             actual:
>             "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
>     
>     Failing input written to testdata/fuzz/FuzzStream/191f25cd0d90d0e0
>     To re-run:
>     go test -run=FuzzStream/191f25cd0d90d0e0
> FAIL
> exit status 1
> FAIL	github.com/bzick/tokenizer	1.489s
> $ cat testdata/fuzz/FuzzStream/191f25cd0d90d0e0
> go test fuzz v1
> string("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA0000")
>  

 _Originally posted by @n-peugnet in [#42](https://github.com/bzick/tokenizer/issues/42#issuecomment-4902135103)_

----

~~By the way, this also asks the question of what to do if the first token is larger than the size of the buffer.~~ EDIT: this seems to be correctly handled but also invalidly triggers the end of the stream.

--- a/stream.go
+++ b/stream.go
@@ -103,13 +103,14 @@ func (s *Stream) GetParsedLength() int {
 // If there is no token, it initiates the parsing of the next chunk of data.
 // If there is no data, the pointer will point to the TokenUndef token.
 func (s *Stream) GoNext() *Stream {
+	// The next token might not have been parsed yet, try to do it now.
+	if s.current.next == nil && s.p != nil {
+		n := s.p.n
+		s.p.parse()
+		s.len += s.p.n - n
+	}
 	if s.current.next != nil {
 		s.current = s.current.next
-		if s.current.next == nil && s.p != nil { // lazy load and parse next data-chunk
-			n := s.p.n
-			s.p.parse()
-			s.len += s.p.n - n
-		}
 		if s.historySize != 0 && s.current.id-s.head.id > s.historySize {
 			t := s.head
 			s.head = s.head.unlink()

--- a/stream.go
+++ b/stream.go
@@ -99,16 +99,20 @@ func (s *Stream) GetParsedLength() int {
 	}
 }
 
-// GoNext moves the stream pointer to the next token.
-// If there is no token, it initiates the parsing of the next chunk of data.
-// If there is no data, the pointer will point to the TokenUndef token.
-func (s *Stream) GoNext() *Stream {
-	// The next token might not have been parsed yet, try to do it now.
-	if s.current.next == nil && s.p != nil {
+// ensureParseNext lazily parses the next chunk of data when the token list ends mid-stream.
+func (s *Stream) ensureParseNext() {
+	if s.current.next == nil && s.current != undefToken && s.p != nil {
 		n := s.p.n
 		s.p.parse()
 		s.len += s.p.n - n
 	}
+}
+
+// GoNext moves the stream pointer to the next token.
+// If there is no token, it initiates the parsing of the next chunk of data.
+// If there is no data, the pointer will point to the TokenUndef token.
+func (s *Stream) GoNext() *Stream {
+	s.ensureParseNext()
 	if s.current.next != nil {
 		s.current = s.current.next
 		if s.historySize != 0 && s.current.id-s.head.id > s.historySize {
@@ -263,6 +267,7 @@ func (s *Stream) PrevToken() *Token {
 // If next token doesn't exist, the method returns TypeUndef token.
 // Do not save a result (Token) into variables — the next token may be changed at any time.
 func (s *Stream) NextToken() *Token {
+	s.ensureParseNext()
 	if s.current.next != nil {
 		return s.current.next
 	}

--- a/stream_test.go
+++ b/stream_test.go
@@ -309,6 +309,27 @@ func TestIssue30(t *testing.T) {
 	}
 }
 
+// TestIssue44 The parser stops parsing the stream after the first token if
+// its size greater or equal to the size of the parser's buffer.
+func TestIssue44(t *testing.T) {
+	parser := New()
+	parser.AllowKeywordSymbols(Underscore, Numbers)
+
+	// stream with first token close to the size of the parser's internal buffer
+	buf := bytes.NewBufferString("this_token_is_exactly_40_characters_long 67")
+	stream := parser.ParseStream(buf, 40)
+	defer stream.Close()
+
+	require.True(t, stream.IsValid())
+	require.True(t, stream.CurrentToken().IsKeyword())
+	require.Equal(t, "this_token_is_exactly_40_characters_long", stream.CurrentToken().ValueString())
+
+	stream.GoNext()
+	require.True(t, stream.IsValid(), "stream should still be valid after first token")
+	require.True(t, stream.CurrentToken().IsInteger())
+	require.Equal(t, int64(67), stream.CurrentToken().ValueInt64())
+}
+
 func TestStreamOverflow(t *testing.T) {
 	parser := New()
 	buf := bytes.NewBuffer([]byte("a b c"))

--- a/stream_test.go
+++ b/stream_test.go
@@ -330,6 +330,25 @@ func TestIssue44(t *testing.T) {
 	require.Equal(t, int64(67), stream.CurrentToken().ValueInt64())
 }
 
+// TestIssue44NextToken NextToken should be able to peek the next token even
+// when the first token's size is grreater or equal to the size of the parser's
+// buffer.
+func TestIssue44NextToken(t *testing.T) {
+	parser := New()
+	parser.AllowKeywordSymbols(Underscore, Numbers)
+
+	// stream with first token close to the size of the parser's internal buffer
+	buf := bytes.NewBufferString("this_token_is_exactly_40_characters_long 67")
+	stream := parser.ParseStream(buf, 40)
+	defer stream.Close()
+
+	require.True(t, stream.IsValid())
+	next := stream.NextToken()
+	require.NotEqual(t, undefToken, next, "NextToken should not return undefToken")
+	require.True(t, next.IsInteger())
+	require.Equal(t, int64(67), next.ValueInt64())
+}
+
 func TestStreamOverflow(t *testing.T) {
 	parser := New()
 	buf := bytes.NewBuffer([]byte("a b c"))

